Deduplicate various steps in CI files

krzysz00 opened the pull request:
- Add most of the flags we use in CI to the CMakeLists.txt

 - Introduce a BUILD_FAT_LIBMLIRMIOPEN to control building the fat MIOpen interface library and avoid building it when it's not needed

 - Move the installation of said library to a separate component, so  now we don't need to run our build tools with root permissions

 - Replace many shell scripts that call cmake and a build tool with calls to the cmake plugin

- Similarly, move PATH changes into the Jenkins configuration

Many of these CI files still contain similar steps (like various
compile/test combinations). Once I have a better sense of what tests
are being run and where, I will be able to combine these files into a
something less redundant and include, for example, conditions for
excluding longer tests on PR-triggered CI.

krzysz00 committed:
Deduplicate various steps in CI files

 - Add most of the flags we use in CI to the CMakeLists.txt

 - Introduce a BUILD_FAT_LIBMLIRMIOPEN to control building the fat
MIOpen interface library and avoid building it when it's not needed

 - Move the installation of said library to a separate component, so
 now we don't need to run our build tools with root permissions

 - Replace many shell scripts that call cmake and a build tool with
 calls to the cmake plugin

- Similarly, move PATH changes into the Jenkins configuration

Many of these CI files still contain similar steps (like various
compile/test combinations). Once I have a better sense of what tests
are being run and where, I will be able to combine these files into a
something less redundant and include, for example, conditions for
excluding longer tests on PR-triggered CI.

krzysz00 committed:
Newlines in bad places break everything

krzysz00 committed:
Missed a steps{} directive

krzysz00 committed:
include/includes change

krzysz00 committed:
Image name is wrong

krzysz00 committed:
Call depenndency installer

krzysz00 committed:
Adjust Docker image label

krzysz00 committed:
Update path setup

krzysz00 committed:
Make MIOpen build actually run in parallel

krzysz00 committed:
SourceDir shouldn't in the thing

krzysz00 committed:
Skip default checkout for test nodes

krzysz00 committed:
Move options block to where it should be

krzysz00 commented:
I'll want to hold off on merging this until #223 lands, since that'll affect the changes I'm making and change the flags I'm supposed to set. However, I'll still take review

krzysz00 committed:
Merge branch 'miopen-dialect' into ci-consolidation-and-image-update

yiqian1 commented on mlir/utils/jenkins/Jenkinsfile.downstream:
```diff
@@ -16,28 +19,12 @@ pipeline {
+                    buildType: 'Release',\
+                    installation: 'InSearchPath',\
+                    steps: [[args: 'check-mlir', withCmake: true]],\
+                    cmakeArgs: '-DMLIR_MIOPEN_DRIVER_ENABLED=1'
```
Also need -DMLIR_MIOPEN_DRIVER_PR_E2E_TEST_ENABLED=1 here.

krzysz00 replied:
`git merge` messed that up, fixed, thanks for catching it

krzysz00 committed:
Put back in a config directive the merge threw out.

jerryyin commented on mlir/tools/mlir-miopen-lib/CMakeLists.txt:
```diff
@@ -75,9 +73,14 @@ if( NOT BUILD_SHARED_LIBS )
-                       COMMAND ${CMAKE_COMMAND} -E copy ${full_output_path} ${install_path}
-                       COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/Miir.h ${header_path}
                        DEPENDS ${output_archive_dummy_file})
+
+    install(FILES ${full_output_path}
+            TYPE LIB
+            COMPONENT libMLIRMIOpen EXCLUDE_FROM_ALL)
+    install(FILES Miir.h
+            TYPE INCLUDE
+            COMPONENT libMLIRMIOpen EXCLUDE_FROM_ALL)
```
[Question] Just want to confirm that there's no change of behavior here. Building the target still install the library?

krzysz00 replied:
There is a change of behavior. Building the target now does *not* install the library, hence the calls in the Jenkins file to `cmake --install`

jerryyin replied:
We synced offline, the new build install sequence is:

```bash
cmake -G "Unix Makefiles" .. -DCMAKE_BUILD_TYPE=Release -DBUILD_FAT_LIBMLIRMIOPEN=1
make -j libMLIRMIOpen
cmake --install . --component libMLIRMIOpen
```

krzysz00 replied:
You don't necessarily need to call for Makefiles - you could also, say, call for Ninja, or just leave `-G` off and let it default to Makefiles

jerryyin replied:
> or just leave -G off and let it default to Makefiles

Thanks for the suggestion

jerryyin commented on mlir/utils/jenkins/Jenkinsfile.downstream:
```diff
@@ -6,6 +6,9 @@ pipeline {
         }
     }
+    environment {
+        PATH="$PATH:/opt/rocm/llvm/bin"
```
I'm not sure this is required to build/check-mlir. I think this is required on `Static Test` stage. If that's the case, could you add a comment to be explicit on which step need this environment?

krzysz00 replied:
I figured  that, since it's used all over the place, this should be placed on the top to avoid duplication (especially since it won't break anything)

jerryyin commented on mlir/utils/jenkins/Jenkinsfile.miopen-test:
```diff
@@ -33,55 +75,17 @@ pipeline {
--DCMAKE_RULE_MESSAGES=OFF
-'''
-                                sh 'cd build; make -j $(nproc) MIOpenDriver'
+                            dir("MIOpen") {
+                                unstash name: "MIOpen-build"
```
This copies the build step package to test stage? Nice!

krzysz00 replied:
Yep.

jerryyin reviewed (commented):
Do you see the possibility to further consolidate everything into 2 jenkinsfile
 - 1 nightly jenkins file
 - 1 PR triggered jenkins file

krzysz00 committed:
Initial test of the one big Jenkinsfile

This Jenkinsfile is parameterized by "nightly" and "canXdlops" -
booleans that should be set to reflect if the job is being run on pull
requests or not (to remove expensive/unneeded tests) and whether the
job is being run in a CI that can run xdlops (that is, if we're on the
private or public CI).

If this file works as planned, the idea is that there'll be four
Jenkins jobs, two on each CI server, which'll run all the relevant
testing. An advantage to this is that a lot of common infrastructure
is factored out, as in the buildProject() function, and that we can
see all our CI jobs in one place. This also eliminates the need to
have many different CI pipeline configurations, simplifying the
administration of the cluster.

yiqian1 commented on mlir/utils/jenkins/Jenkinsfile.downstream:
```diff
@@ -16,65 +19,37 @@ pipeline {
+                    buildDir: 'build',\
+                    buildType: 'Release',\
+                    installation: 'InSearchPath',\
+                    steps: [[args: 'check-mlir', withCmake: true]],\
```
Should this be --target check-mlir?
I'm sure the tools were built but tests were not run. Please check the console outputs.

krzysz00 replied:
Could be, and I might remove the withCmake since it really isn't needed

krzysz00 commented:
Re consilidaiton: I'm currently writing a Jenkinsfile that merges everything into one file that's parametrized by public/private and nightly/PR

krzysz00 committed:
Remove a bunch of withCmake calls that aren't needed

krzysz00 committed:
Missing close brace

krzysz00 committed:
Correct the syntax for sequential-in-parallel

krzysz00 committed:
Clean up some syntax issues, hopefully

krzysz00 committed:
Remove large Jenkinsfile - development moving to a new branch

krzysz00 committed:
Fix test failures, especially when miopen driver isn't built

1. Disable a standalone test that's broken and irrelevant to us
2. Make certain miopen-related directories tests conditional on the
miopen driver being built
3. Disable some MIOpen dialect tests that use the driver when the
driver isn't compiled

krzysz00 committed:
Add in sourceDir for LLVM build

jerryyin commented on mlir/test/Examples/standalone/test.toy:
```diff
@@ -1,3 +1,5 @@
+# This doesn't substitute correctly right now
+# XFAIL: *
```
We don't care about this directory. However, do you see the need of filing a ticket against ourselves in fixing the cmake version?

yiqian1 replied:
This test used to pass. What caused it to fail?

krzysz00 replied:
It's not clear

jerryyin replied:
@krzysz00 I remember you mentioned that it is related with cmake. What's the last passing cmake version?

krzysz00 replied:
I can try and take a look at it

whchung replied:
Need to get to the bottom of this.

jerryyin commented on mlir/utils/jenkins/Jenkinsfile.upstream:
```diff
@@ -21,6 +21,7 @@ pipeline {
                 git 'https://github.com/llvm/llvm-project'
                 cmakeBuild generator: 'Ninja',\
                     buildDir: 'build',\
+                    sourceDir: 'llvm',\
```
Why does this `sourceDir` matter? Why is it not `mlir`?

krzysz00 replied:
`sourceDir` here is the directory containing the CMakeLists.txt, and, in the previous Jenkinsfile, we built from `llvm` (and that's what our current setup also does - it builds `llvm`, which then goes on to include `mlir` under `tools/`)

jerryyin replied:
I'm not convinced this does anything. Taking a look back, I think that `../llvm` is added by mistake and you can build with `..` all the same - We do also have a `CMakeLists.txt` in the root directory.

Since we are not using this jenkinsfile anyways, I'm fine with you keep this as-is.

krzysz00 replied:
If you look at the context, this isn't our build - this is a build of LLVM from https://github.com/llvm/llvm-project - which doesn't have a CMakeLists.txt in the root.

However, if I misunderstood the setup and the idea is to test our code but with upstream's files, then I can change it.

jerryyin replied:
> this is a build of LLVM from

You get the idea right. Though, I still remember when I used to compile the tip of upstream llvm and don't need to do `../llvm` in the configuration step. 

I think this is an unimportant implementation detail that can always get verified easily when we start to build upstream tip again. I'm marking this as resolved.

jerryyin reviewed (commented):
LGTM. You have my approval. 

@yiqian1 Due to the magnitude of changes in this PR, please help take a second look. You are more familiar with each individual jenkinsfile. Approve and merge if you are also okay with it.

krzysz00 committed:
Maybe it's nested docker() nodes that cause issues - will revert if this isn't it.

krzysz00 committed:
Move scripts into MIOpen test environment

krzysz00 committed:
Switch image to one with 'parallel' in it

krzysz00 commented:
Good news, I found a workaround for the MIOpen test, see http://10.236.106.97:8080/blue/organizations/jenkins/MLLIR-MIopen-Resnet50-refactor-dev/detail/MLLIR-MIopen-Resnet50-refactor-dev/35/pipeline/526 (turns out the issue is that we can't have a docker agent definition nested inside another docker agent definition)

jerryyin commented on mlir/utils/jenkins/Jenkinsfile.miopen-test:
```diff
@@ -64,7 +64,7 @@ mlir/utils/jenkins/miopen-tests/**\
                 agent {
                     docker {
-                        image 'rocm/mlir:rocm4.1-latest'
+                        image 'rocm/mlir:latest'
```
'rocm/mlir:rocm4.1-latest' should be equivalent with 'rocm/mlir:latest', since ROCm 4.1 is what we got.

krzysz00 replied:
For unknown reasons, rocm4.1-latest didn't have `parallel` in it, maybe it's because we ROCm 4.2 has been released so the update didn't get given that tag?

jerryyin replied:
> maybe it's because we ROCm 4.2 has been released

As long as we haven't bumped to ROCm 4.2 in the dockerfile, it won't happen automatically.

I took a look a second ago and see `rocm/mlir:latest` and `rocm4.1-latest` having the exact same commit hash. Can you try to pull again, I suspect you have a out-dated image.

krzysz00 replied:
I'm not sure how to get Jenkins to re-pull

jerryyin replied:
I did some googling and find [this image](https://i.stack.imgur.com/ZP0vN.png), [here](https://stackoverflow.com/questions/46327298/jenkins-always-pull-docker-image-from-a-declarative-pipeline-agent-definition). Could you verify this? 

`alwaysPull true`

I think the idea is to force jenkins pull the latest docker instead of using the local out-of-dated cached image.

krzysz00 replied:
We may not want to *always* have that on, but it's probably a good idea since we've recently made changes to our Docker images

krzysz00 replied:
`alwaysPull true` does what we want - it always pulls, but `pull` itself won't redownload if nothing's changed

We may just want to check that we're not building new images too often.

jerryyin replied:
Thanks for confirming. 

> pull itself won't redownload if nothing's changed

That's pretty reasonable behavior.

> We may just want to check that we're not building new images too often.

Agreed, but I think we are already doing the best we can. Besides, the overhead of downloading a fresh image isn't that much so I won't worry about it now.

jerryyin commented:
> I found a workaround for the MIOpen test

Good job!

yiqian1 commented:
LGTM. Just wonder what caused the standalone test to fail.

krzysz00 committed:
Try out alwaysPull, see what it does

jerryyin reviewed (approved):
Solid work! Thanks

whchung commented:
@jerryyin / @krzysz00 This PR needs to be reverted as it breaks `check-mlir`.

whchung commented:
Filed #233 to revert this PR.

whchung commented on mlir/test/Examples/standalone/test.toy:
```diff
@@ -1,3 +1,5 @@
+# This doesn't substitute correctly right now
+# XFAIL: *
```
Why is this needed? It's breaking `check-mlir`
